### PR TITLE
Fix CHIRRTL bugs

### PR DESCRIPTION
**Pull request by donggyukim**
Addresses #325, #332, and overrides #326.

It's a bit surprising that it has been working with wrong type inferences for CHIRRTL.

### CODE REVIEW

**Comment by azidar** [comment]
The types get over-written later on in the InferTypes pass, so that probably cleared out the incorrectly inferred types.

File: src/main/scala/firrtl/passes/InferTypes.scala [hunk 1]
```diff
@@ -103,8 +103,9 @@ object CInferTypes extends Pass {
         types(sx.name) = sx.tpe
         sx
       case sx: DefNode =>
-        types(sx.name) = sx.value.tpe
-        sx
+        val sxx = (sx map infer_types_e(types)).asInstanceOf[DefNode]
+        types(sxx.name) = sxx.value.tpe
+        sxx
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/InferTypes.scala]: Why does `DefRegister` call `infer_types_e` on it's child `Expressions` but `DefWire` does not?

Comment (donggyukim) [inline src/main/scala/firrtl/passes/InferTypes.scala]: I have no idea, but `InferTypes` is written in this manner...

Comment (jackkoenig) [inline src/main/scala/firrtl/passes/InferTypes.scala]: @azidar?

Comment (donggyukim) [inline src/main/scala/firrtl/passes/InferTypes.scala]: I see, `DefWire` has no expressions.

Comment (jackkoenig) [inline src/main/scala/firrtl/passes/InferTypes.scala]: Haha whoops, duh

**Pull request merged**